### PR TITLE
[POAE7-2682] LLVMJIT TargetMachine Configuration

### PR DESCRIPTION
**Pull request by bigPYJ1151**
### What changes were proposed in this pull request?

Use ```TargetMachine``` for more precise compile configuration control.

### Why are the changes needed?

Performance evaluation.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unnecessary

### Which label does this PR belong to?

Infra

Pls take a review @Ma-Jian1 @winningsix

### CODE REVIEW

File: cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h [hunk 1]
```diff
@@ -26,6 +26,21 @@
...
+  // TBD other optimizeLevel to be added
+};
+
+// compilation config info
+struct CompilationOptions {
+  LLVMJITOptimizeLevel optimize_level = LLVMJITOptimizeLevel::RELEASE;
+  bool aggresive_jit_compile = true;
+  bool dump_ir = false;
+  bool enable_avx = false;
+  bool enable_avx512 = false;
```
Comment (Ma-Jian1) [inline cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h:41]: maybe we should enable avx automatically if avx512 is enabled?

**Review by xieqi — APPROVED** [review]

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 2]
```diff
@@ -30,12 +32,52 @@ LLVMJITEngine::~LLVMJITEngine() {
...
+  }
+  return target;
+}();
+
+static llvm::StringMap<bool> host_supported_features = []() {
+  llvm::StringMap<bool> features;
+  if (!llvm::sys::getHostCPUFeatures(features)) {
+    LOG(FATAL) << "Unable to get host supported features.";
+  }
+  // TBD (bigPYJ1151): whether need to filter unused instruction sets.
```
Comment (winningsix) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:75]: Q: what's the cost of adding unused IS?

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 3]
```diff
@@ -44,18 +86,53 @@ static llvm::TargetOptions buildTargetOptions() {
...
+llvm::TargetMachine* LLVMJITEngineBuilder::buildTargetMachine(
+    const CompilationOptions& co) {
+  return host_target->createTargetMachine(
+      process_triple,
+      process_name,
+      buildTargetFeatures(co).getString(),
+      buildTargetOptions(),
+      llvm::None,
+      llvm::None,
+      co.aggresive_jit_compile ? llvm::CodeGenOpt::Aggressive : llvm::CodeGenOpt::Default,
```
Comment (winningsix) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:118]: typo: ```aggresive_jit_compile``` -> ```aggressive_jit_compile```

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 4]
```diff
@@ -44,18 +86,53 @@ static llvm::TargetOptions buildTargetOptions() {
...
+  llvm::StringMap<bool> features_copy(host_supported_features);
+  auto switch_inst_set = [&features_copy](auto&& sets, bool flag) {
+    for (auto feature : sets) {
+      if (auto iter = features_copy.find(feature); features_copy.end() != iter) {
+        iter->second &= flag;
+      }
+    }
+  };
+
+  switch_inst_set(avx256_inst_sets, co.enable_avx);
```
Comment (winningsix) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:99]: ```enable_avx``` -> ```enable_avx2```?

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h [hunk 5]
```diff
@@ -26,6 +26,21 @@
...
+  RELEASE
+  // TBD other optimizeLevel to be added
+};
+
+// compilation config info
+struct CompilationOptions {
+  LLVMJITOptimizeLevel optimize_level = LLVMJITOptimizeLevel::RELEASE;
+  bool aggresive_jit_compile = true;
+  bool dump_ir = false;
+  bool enable_avx = false;
```
Comment (winningsix) [inline src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.h:40]: AVX2 minimal required? ```enable_avx2 = true```

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp [hunk 6]
```diff
@@ -30,12 +32,52 @@ LLVMJITEngine::~LLVMJITEngine() {
...
+                                         "avx512vpopcntdq",
+                                         "avx512f",
+                                         "avx512bw",
+                                         "avx512vbmi2",
+                                         "avx512vl",
+                                         "avx512cd",
+                                         "avx512vbmi",
+                                         "avx512bf16",
+                                         "avx512dq",
+                                         "avx512pf"};
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:50]: Yes, we need to add all necessary IS to this list. It is necessary to figure out purpose of different sub-IS and whether add it into the list.

File: src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp (hunk 2 above)
Comment (bigPYJ1151) [reply src/cider/exec/nextgen/jitlib/llvmjit/LLVMJITEngine.cpp:75]: Not sure, maybe affect compile time.

**Review by winningsix — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Pull request merged**